### PR TITLE
sync prometheus ext tests running in parallel

### PR DESCRIPTION
**Pull request by gabemontero**
Fixes https://github.com/openshift/origin/issues/17693

@bparees @smarterclayton @mfojtik fyi

### CODE REVIEW

**Comment by bparees** [comment]
/lgtm
thanks @gabemontero

**Comment by gabemontero** [comment]
conformance install failure flake https://github.com/openshift/origin/issues/17605

/test extended_conformance_install

File: test/extended/prometheus/prometheus.go [hunk 1]
```diff
@@ -292,19 +293,25 @@ func waitForServiceAccountInNamespace(c clientset.Interface, ns, serviceAccountN
...
+	processTemplate bool
+	mux             sync.Mutex
+)
+
 func bringUpPrometheusFromTemplate(oc *exutil.CLI) (ns, host, bearerToken string, statsPort int) {
 	ns = oc.KubeFramework().Namespace.Name
 	host = "prometheus.kube-system.svc"
 	statsPort = 443
-	mustCreate := false
+	mux.Lock()
```
Comment (smarterclayton) [inline test/extended/prometheus/prometheus.go]: Won't work, other jobs are run in separate processes.

Comment (gabemontero) [reply test/extended/prometheus/prometheus.go]: Ah ... gotcha.

So seems like we need to interpret the error, and if it is "already exists", do not abort the test case.

I'll circle back and give that a go (unless some other direction is supplied).

Comment (gabemontero) [reply test/extended/prometheus/prometheus.go]: I'll see if I can remove the lgtm label

/hold

Comment (gabemontero) [reply test/extended/prometheus/prometheus.go]: appears I can :-)

**Comment by gabemontero** [comment]
adjustment pushed ... fyi, totally removed the mutex (seemed unnecessary given the multi-process basis)

File: test/extended/prometheus/prometheus.go [hunk 2]
```diff
@@ -310,7 +310,11 @@ func bringUpPrometheusFromTemplate(oc *exutil.CLI) (ns, host, bearerToken string
 		stdout, _, err := oc.WithoutNamespace().Run("process").Args("-f", configPath).Outputs()
 		o.Expect(err).NotTo(o.HaveOccurred())
 		err = oc.WithoutNamespace().AsAdmin().Run("create").Args("-f", "-").InputString(stdout).Execute()
-		o.Expect(err).NotTo(o.HaveOccurred())
+		// still check if it exists, as the prior not found check above may have been in a race with a test
+		// running in another process
+		if !kapierrs.IsAlreadyExists(err) {
```
Comment (bparees) [inline test/extended/prometheus/prometheus.go]: you're not making an api call here, you're effectively shelling out to invoke "oc create -f", so this error check is not going to work.

Comment (bparees) [reply test/extended/prometheus/prometheus.go]: you may have to literally scan the output from the command.  Or just log the error w/o failing the test (so if the test ultimately fails, we can at least go back and discover the error that occurred)

Comment (gabemontero) [reply test/extended/prometheus/prometheus.go]: I'll go with the latter ... the existing use of `IsNotFound` presumably also falls into this bucket

Comment (bparees) [reply test/extended/prometheus/prometheus.go]: The existing call is not using the oc client:
```
if _, err := oc.AdminKubeClient().Apps().StatefulSets("kube-system").Get("prometheus", metav1.GetOptions{}); err != nil {
if !kapierrs.IsNotFound(err) {
```

it's using the KubeClient (aka the API client)

so it's fine.

You could consider doing the same thing here.

**Comment by gabemontero** [comment]
the error in https://openshift-gce-devel.appspot.com/build/origin-ci-test/pr-logs/pull/17717/test_pull_request_origin_extended_conformance_install/3954/ 
appears to be a problem starting builds because of the space / devmapper issue @coreydaley is tracking

File: test/extended/prometheus/prometheus.go (hunk 2 above)
Comment (gabemontero) [reply test/extended/prometheus/prometheus.go]: I've been exploring the various client entryways in cli.go but am not finding a verb that allows me to "process" the template

I'll look a bit more but unless you have a precise pointer, thinking of falling back to parsing output.

And thanks for the clarification around the not found path.

Comment (bparees) [reply test/extended/prometheus/prometheus.go]: you don't have to change the process logic, just the create logic.

Comment (bparees) [reply test/extended/prometheus/prometheus.go]: but honestly this should fail so rarely i'd be fine w/ you just logging the error so if the test fails further down, we have the error on record.

**Comment by gabemontero** [comment]
update pushed @bparees

**Comment by gabemontero** [comment]
rebase pushed ... @bparees please re-review / re-post the good to merge comment at your convenience

**Comment by gabemontero** [comment]
Opened https://github.com/openshift/origin/issues/17811 for the cmd failure.

**Comment by gabemontero** [comment]
conformance gce was my flake https://github.com/openshift/origin/issues/17694

hopefully the debug data PR merged in time for it to capture data for this one.

**Comment by gabemontero** [comment]
conformance install was pre-test, env failures in origin prereqs and install origin

**Comment by gabemontero** [comment]
conformance install failure was flake https://github.com/openshift/origin/issues/17556

**Comment by gabemontero** [comment]
opened and taking flake https://github.com/openshift/origin/issues/17836

**Pull request merged**